### PR TITLE
CLI tools to make it easier to work with SealedSecrets

### PR DESCRIPTION
**Pull request by joakimnordling**
This PR adds three new invoke commands:

1. `poetry run invoke get-master-key --env <env>` to make it easy to back up the `master.key` for an environment.
2. `poetry run invoke unseal-secrets --env <env>` to decrypt all the secrets using the `master.key` (retrieving it if it's not present) and base64 decode them so they are easy to view and modify.
3. `poetry run invoke seal-secrets --env <env>` to base64 encode the secrets and then seal them.

The critical files (`*.unsealed-secrets.yaml` and `master.key`) have been added to .gitignore and the README.md has been updated with more or less the same info as stated above.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -473,6 +473,37 @@ You might also want to back-up your master key, but do NOT store it in
...
+If you have the `master.key` file for an environment (or sufficient privileges to retrieve it from the cluster),
+you can use the following command to unseal the secrets to a readable and editable format:
+
+```bash
+poetry run invoke unseal-secrets --env <env>
+```
+This will convert each of the files in `envs/<env>/secrets/*.yaml` to a corresponding
+`*.unsealed-secrets.yaml` file. This will also base64 decode the file, so you can see and edit the actual values easily.
+
+N.B. Make sure to not under any circumstance commit the `*.unsealed-secrets.yaml` or `master.key` files!
```
Comment (lietu) [inline README.md]: `N.B.` is not a very widely known notation :) How about just `NOTE:`

File: devops/tasks.py [hunk 2]
```diff
@@ -209,3 +224,198 @@ def release(
...
+            "yaml",
+        ]
+    )
+
+    content = result.stdout.decode(encoding="utf-8")
+    output_file = master_key_path(env=env)
+
+    logger.info(f"Saving master key to {output_file}")
+
+    with output_file.open("w", encoding="utf-8") as f:
```
Comment (lietu) [inline devops/tasks.py]: `output_file` is a `Path`, so `output_file.write_text(content, encoding="utf-8")`?

File: devops/tasks.py [hunk 3]
```diff
@@ -209,3 +224,198 @@ def release(
...
+    ]
+
+    label(logger.info, f"Unsealing secrets for {env}")
+
+    for input_file in sealed_secret_files:
+        output_file = input_file.with_name(input_file.stem + UNSEALED_SECRETS_EXTENSION)
+
+        logger.info(f"Unsealing {input_file} to {output_file}")
+
+        with input_file.open("r", encoding="utf-8") as f:
```
Comment (lietu) [inline devops/tasks.py]: `Path.read_text(encoding="utf-8")`?

File: devops/tasks.py [hunk 4]
```diff
@@ -209,3 +224,198 @@ def release(
...
+            content = f.read()
+
+        content = kube_unseal(content, master_key)
+        content = base64_decode_secrets(content)
+
+        with output_file.open("w", encoding="utf-8") as f:
```
Comment (lietu) [inline devops/tasks.py]: `output_file.write_text(content, encoding="utf-8")`?

File: devops/tasks.py [hunk 5]
```diff
@@ -209,3 +224,198 @@ def release(
...
+
+    label(logger.info, f"Sealing secrets for {env}")
+
+    for input_file in unsealed_secret_files:
+        output_file_name = input_file.name[: -len(UNSEALED_SECRETS_EXTENSION)] + ".yaml"
+        output_file = input_file.with_name(output_file_name)
+
+        logger.info(f"Sealing {input_file} as {output_file}")
+
+        with input_file.open("r", encoding="utf-8") as f:
```
Comment (lietu) [inline devops/tasks.py]: Same as above

File: devops/tasks.py [hunk 6]
```diff
@@ -209,3 +224,198 @@ def release(
...
+            content = f.read()
+
+        content = base64_encode_secrets(content)
+        content = kube_seal(content, cert=secrets_pem)
+
+        with output_file.open("w", encoding="utf-8") as f:
```
Comment (lietu) [inline devops/tasks.py]: Same as above

File: devops/tasks.py [hunk 7]
```diff
@@ -209,3 +224,198 @@ def release(
...
+        with input_file.open("r", encoding="utf-8") as f:
+            content = f.read()
+
+        content = kube_unseal(content, master_key)
+        content = base64_decode_secrets(content)
+
+        with output_file.open("w", encoding="utf-8") as f:
+            f.write(content)
+
+        input_file.unlink()
```
Comment (lietu) [inline devops/tasks.py]: Do we really want to delete the sealed files?

File: devops/tasks.py [hunk 8]
```diff
@@ -209,3 +224,198 @@ def release(
...
+        with input_file.open("r", encoding="utf-8") as f:
+            content = f.read()
+
+        content = base64_encode_secrets(content)
+        content = kube_seal(content, cert=secrets_pem)
+
+        with output_file.open("w", encoding="utf-8") as f:
+            f.write(content)
+
+        input_file.unlink()
```
Comment (lietu) [inline devops/tasks.py]: In this case, deleting the unsealed files makes more sense though :)

File: devops/tasks.py [hunk 9]
```diff
@@ -209,3 +224,198 @@ def release(
...
+    secrets = yaml.load(content)
+
+    data = secrets["data"]
+    for key, value in data.items():
+        if value is not None:
+            value = base64.b64decode(value.encode("utf-8")).decode("utf-8")
+            if "\n" in value:
+                # If there's a line break in the value we want to dump it using
+                # the literal syntax
+                value = LiteralScalarString(value)
```
Comment (lietu) [inline devops/tasks.py]: Not quite sure what this means - could you add an example?

File: devops/lib/utils.py [hunk 10]
```diff
@@ -57,6 +65,7 @@ def run(
     :param dict env:
     :param bool stream: If the output should be streamed instead of captured
     :param float timeout: Seconds to wait before failing
+    :param bytes input: Data to be sent to the child process
```
Comment (lietu) [inline devops/lib/utils.py]: Might be a good idea to mention `via stdin`

File: README.md (hunk 1 above)
Comment (joakimnordling) [reply README.md]: Changed according to suggestion.

File: devops/tasks.py (hunk 2 above)
Comment (joakimnordling) [reply devops/tasks.py]: Nice, somehow has missed that feature existed. Changes all of them.

File: devops/tasks.py (hunk 7 above)
Comment (joakimnordling) [reply devops/tasks.py]: Probably better to leave the files than unexpectedly delete them. The .gitgnore should hopefully avoid accidentally committing the files to the repo.

File: devops/tasks.py (hunk 8 above)
Comment (joakimnordling) [reply devops/tasks.py]: To be consistent with above, also changed this to not delete. It's also possible that you don't have access to the master.key, in which case deleting the file could remove your only copy of the info, so better to leave the files.

File: devops/tasks.py (hunk 9 above)
Comment (joakimnordling) [reply devops/tasks.py]: Improved the explanation. Better like this, or still ideas on how to improve it?

Comment (lietu) [reply devops/tasks.py]: it's perfect now

**Pull request merged**